Add id field to the changeset items structure.

UnderGreen opened the pull request:
Adding ID fields to use with `rig upsert/delete` commands makes them
some sort of idempotent.

UnderGreen committed:
Add id field to the changeset items structure.

Adding ID fields to use with `rig upsert/delete` commands makes them
some sort of idempotent.

r0mant commented on changeset.go:
```diff
@@ -109,63 +109,94 @@ func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetNa
 	}
-	switch kind.Kind {
+
+	comp, err := cs.isOpCompleted(ctx, tr, id)
```
Nit: Let's not shorten words unnecessarily: `completed, err := cs.isOperationCompleted(ctx, tr, id)`.

r0mant commented on changeset.go:
```diff
@@ -109,63 +109,94 @@ func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetNa
+		log := log.WithFields(log.Fields{
+			"cs": tr.String(),
+		})
+		log.Infof("%s '%s' is already updated.", headers.TypeMeta.Kind, headers.ObjectMeta.Name)
```
Nit: I'd add kind and name to the fields as well:
```go
log.WithFields(log.Fields{
  "cs": tr.String(),
  "kind": ...,
  "name": ...,
}).Info("Resource already upserted.")
```

r0mant commented on changeset.go:
```diff
@@ -109,63 +109,94 @@ func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetNa
+		return nil
+	}
+
+	if err != nil {
```
Not sure I understand this logic. So if the resource hasn't been upserted yet, shouldn't we go ahead and upsert it below? Why are we calling the status here? And what to do if the status fails and returns that "operation not completed" error below?

UnderGreen replied:
We can't upsert it again, because the operation was already started. If you start it for the second time you will lose original resource. And technically I have no solution yet how to handle error in status. Probably it should be the operation revert + upsert again. But it will not help in cases when a new resource can't start because of taints/node disk/cpu/memory pressures and etc. Looks like error should be handled manually.

UnderGreen replied:
The scenario which I am trying to solve here:
1. I am running `rig upsert` for specific resource. It finished fine, resource is update.
2. Later my script failed somewhere and I need to restart the whole script - typical case for update hook, right?
3. I run it for the second time and if I do upsert again my original resource will be overwritten by the resource which I updated in step one. I lost the ability to revert everything back because I don't have original resources anymore.

My approach in this PR:
1. Update resource if `rig upsert ` for that specific resource(Kind, Name, Namespace) was invoked for the first time.
2. Skip the update for the all next invocation for the same resource.

That will allow me to keep original resources and I can rollback if needed.

r0mant commented on changeset.go:
```diff
@@ -109,63 +109,94 @@ func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetNa
 }
 
+// isOpCompleted checks whether operation is completed in the context of a given changeset
+func (cs *Changeset) isOpCompleted(ctx context.Context, tr *ChangesetResource, id string) (bool, error) {
```
Nit: Like mentioned above, I'd rename to `isOperationComplete` for clarity.

r0mant commented on changeset.go:
```diff
@@ -188,6 +219,10 @@ func (cs *Changeset) Status(ctx context.Context, changesetNamespace, changesetNa
 
 	return retry(ctx, retryAttempts, retryPeriod, func() error {
 		for _, op := range tr.Spec.Items {
+			if id != "" && op.ID != id {
```
What does this check for?

r0mant replied:
If the intent here is to differentiate b/w checking the status of a particular operation vs the entire changeset, then I'd do something like this:

```go
const AllOperations = "*"

func (cs *Changeset) Status(ctx context.Context, chNamespace, chName, opID string, ...) ... {
  ...
  return retry(ctx, retryAttempts, retryPeriod, func() error {
    for _, op := range tr.Spec.Items {
      if opID != AllOperations && opID != op.ID {
        continue
      }
    }
  }
}

// To check overall changeset status.
cs.Status(ctx, namespace, name, AllOperations, ...)
```

Or better yet, I'd see if it's possible to have a separate method for checking particular operation status.

a-palchikov replied:
Let's add the Status API for the specific operation.

r0mant commented on changeset.go:
```diff
@@ -231,35 +266,42 @@ func (cs *Changeset) DeleteResource(ctx context.Context, changesetNamespace, cha
+
+	comp, err := cs.isOpCompleted(ctx, tr, id)
+	if comp {
+		log.Infof("%s '%s' is already deleted.", resource.Kind, resource.Name)
```
Nit: Use log fields for kind/name like above.

r0mant commented on tool/rig/main.go:
```diff
@@ -48,18 +49,20 @@ func run(quiet *bool) error {
 		cupsert          = app.Command("upsert", "Upsert resources in the context of a changeset")
 		cupsertChangeset = Ref(cupsert.Flag("changeset", "name of the changeset").Short('c').Envar(changesetEnvVar).Required())
 		cupsertFile      = cupsert.Flag("file", "file with new resource spec").Short('f').Required().String()
+		cupsertID        = cupsert.Flag("id", "ID of a operation in changeset").String()
```
"ID of a changeset operation". Same below.

r0mant commented on tool/rig/main.go:
```diff
@@ -119,21 +123,29 @@ func run(quiet *bool) error {
-		return deleteResource(ctx, client, config, *namespace, *cdeleteChangeset, *cdeleteResourceNamespace, *cdeleteResource, *cdeleteCascade, *cdeleteForce)
+		if *cdeleteID == "" {
+			// Concatenate all arguments and take MD5 from it as an ID
+			*cdeleteID = fmt.Sprintf("%x", md5.Sum([]byte(strings.Join(os.Args[1:], ""))))
```
Couple of thoughts here:

* I'd move this into a helper function and used here and above so the logic is not duplicated.
* I think that this way of calculating the ID has a couple of downsides:
  * It depends on the order of the command-line.
  * It only takes into account CLI args/flags. IIRC changeset ID is usually taken from an env var so it won't be accounted for here.

What if instead taking the whole command-line, we would auto-calculate ID of the operation based on a number of stable parameters: changeset name + operation type (upsert/delete) + resource namespace + resource name? This should uniquely identify a particular operation I think.

r0mant commented on tool/rig/main.go:
```diff
@@ -119,21 +123,29 @@ func run(quiet *bool) error {
 		return freeze(ctx, client, config, *namespace, *cfreezeChangeset)
 	case cupsertConfigMap.FullCommand():
-		return upsertConfigMap(ctx, client, config, *namespace, *cupsertConfigMapChangeset, *cupsertConfigMapName, *cupsertConfigMapNamespace, *cupsertConfigMapFiles, *cupsertConfigMapLiterals)
+		return upsertConfigMap(ctx, client, config, *namespace, *cupsertConfigMapChangeset, *cupsertConfigMapName, *cupsertConfigMapNamespace, *cupsertConfigMapID, *cupsertConfigMapFiles, *cupsertConfigMapLiterals)
```
Why isn't ID auto-calculated here as well?

r0mant commented on tool/rig/main.go:
```diff
@@ -289,7 +302,7 @@ func status(ctx context.Context, client *kubernetes.Clientset, config *rest.Conf
 			return trace.Wrap(err)
 		}
-		err = cs.Status(ctx, namespace, resource.Name, retryAttempts, retryPeriod)
+		err = cs.Status(ctx, namespace, resource.Name, "", retryAttempts, retryPeriod)
```
Why is this passed empty here?

r0mant commented on changeset.go:
```diff
@@ -109,63 +109,94 @@ func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetNa
+			"cs": tr.String(),
+		})
+		log.Infof("%s '%s' is already updated.", headers.TypeMeta.Kind, headers.ObjectMeta.Name)
+		return nil
```
Should we allow `--force` flag to overrule the "already completed" status and upsert anyway? Otherwise there seems to be no way to force-retry. Same for delete.

UnderGreen replied:
As I mentioned in one of my comments you can't just retry operation without losing original resource.

a-palchikov commented on changeset.go:
```diff
@@ -231,35 +266,42 @@ func (cs *Changeset) DeleteResource(ctx context.Context, changesetNamespace, cha
 	switch resource.Kind {
 	case KindDaemonSet:
-		return cs.deleteDaemonSet(ctx, tr, resourceNamespace, resource.Name, cascade)
+		return cs.deleteDaemonSet(ctx, tr, resourceNamespace, id, resource.Name, cascade)
```
This is becoming unmanageable with the proliferation of arguments.
Let's move them to a type similar to:
```go
type deleteResource struct {
	cascade bool
	resource metav1.ObjectMeta
	operationID string
}
```

a-palchikov commented on changeset.go:
```diff
@@ -97,7 +97,7 @@ type Changeset struct {
 
 // Upsert upserts resource in a context of a changeset
-func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetName string, data []byte) error {
+func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetName, id string, data []byte) error {
```
I also think we can do better if we hid `changesetNamespace` and `changesetName` inside of `Changeset`.

a-palchikov commented:
My feeling is that with `resource.UID` already in place there's really no need for this new operation ID.
Besides, I don't know how that works for a set of resources since a single `ChangesetItem` maps to a single resource, but the ID seems to be generated for the whole group (?)

What do you guys think?

UnderGreen commented:
ID is generated per item in changeset. With current implementation of setting UID from the resource itself in idempotency. The second time you run `rig upsert` for the same resource it will be upserted again, as far as I understand the code.

UnderGreen commented on spec.go:
```diff
@@ -57,6 +57,8 @@ type ChangesetSpec struct {
 }
 
 type ChangesetItem struct {
+	// ID represents unique identification for changing operation
```
How to get original resource UIDs from second invocation of rig delete/upsert in that case?

UnderGreen committed:
Generate uuid for new resource

UnderGreen committed:
Addressing review comments

UnderGreen commented:
@a-palchikov, @r0mant 
I changed the logic of checking idempotency for resources here.
No changes from the original branch for `rig upsert`. We have `--force` flag, make no sense to check if the previous operation was completed or not.
For `rig upsert`: I am checking if an item in changeset contains the same values for Kind, Name, and Namespace, which are unique set per K8s cluster. `--force` flag adds the ability to skip the completeness check and upsert again.
For `rig revert`: Added `uid` flag to revert only one operation in changeset and delete that item(because an item with not `in-progress` status can lead to other problems).
Updated formatMeta function and small changes in log lines for all resources.

a-palchikov commented on changeset.go:
```diff
@@ -129,65 +133,83 @@ func (cs *Changeset) upsertResource(ctx context.Context, changesetNamespace, cha
+			return false, trace.Wrap(err)
+		}
+
+		if info.To != nil &&
```
```go
if info.equals(headers) ...
```
nit: Why is the variable called `headers` - does it describe multiple resources?

a-palchikov commented on changeset.go:
```diff
@@ -254,61 +272,59 @@ func (cs *Changeset) Status(ctx context.Context, changesetNamespace, changesetNa
-	if comp {
-		log.Infof("%s '%s' is already deleted.", resource.Kind, resource.Name)
-		return nil
+	dr := deleteResource{
```
`resource` or `res` and `Ref` parameter could be `resourceRef` or just `ref`

a-palchikov commented on changeset.go:
```diff
@@ -361,6 +382,31 @@ func (cs *Changeset) Revert(ctx context.Context, changesetNamespace, changesetNa
+		}
+		return nil
+	}
+	return trace.NotFound("item with %s not found", uid)
```
nit: `with UID %s ...`

a-palchikov commented on changeset.go:
```diff
@@ -1338,28 +1382,27 @@ func (cs *Changeset) withUpsertOp(ctx context.Context, tr *ChangesetResource, ol
 	}
-	tr.Spec.Items[len(tr.Spec.Items)-1].Status = OpStatusCompleted
-	return cs.update(tr)
+	ur.resource.Spec.Items[len(ur.resource.Spec.Items)-1].Status = OpStatusCompleted
```
isn't this the same as setting `item.Status = OpStatusCompleted` before appending it? Seems easier that way

a-palchikov commented on changeset.go:
```diff
@@ -1338,28 +1382,27 @@ func (cs *Changeset) withUpsertOp(ctx context.Context, tr *ChangesetResource, ol
 	}
 	log := log.WithFields(log.Fields{
-		"cs":  tr.String(),
-		"job": fmt.Sprintf("%v/%v", job.Namespace, job.Name),
```
I'd leave the resource in the logger:
```go
	...
	"job": formatMeta(job.ObjectMeta, job.TypeMeta),
```

a-palchikov commented on changeset.go:
```diff
@@ -1323,7 +1368,6 @@ func (cs *Changeset) withUpsertOp(ctx context.Context, tr *ChangesetResource, ol
 		Status:            OpStatusCreated,
-		ID:                id,
 	}
 	newUUID, err := uuid.NewRandom()
```
If this is randomly chosen, how will `rig revert --uid=<uid>` be used?
Also, if the upsert is for a resource being read from a file, `UID` will initially be empty but `old` is not - hence it will reset it to empty string inside the `if` block won't it?

UnderGreen replied:
For the first question. `kubectl get changesets -o yaml` to show all operations and their UIDs. Or we can add command like `rig show` to provide the same output.

UnderGreen commented on changeset.go:
```diff
@@ -97,7 +101,7 @@ type Changeset struct {
 
 // Upsert upserts resource in a context of a changeset
-func (cs *Changeset) Upsert(ctx context.Context, changesetNamespace, changesetName, id string, data []byte) error {
+func (cs *Changeset) Upsert(ctx context.Context, data []byte, force bool) error {
```
Yeah. Probably we need to split the command into an update and upsert. For the update force will make sense.
The point of adding force was to make an ability to update the resource even if it was already updated. Otherwise, if you call upsert for the second time you intentionally or non-intentionally rewrites the old one(original) resource, which makes revert impossible.

r0mant commented:
We chose a simpler approach for now, closing.
